File: frontend/src/lib/health.ts
/**
 * The health index, as a list of named terms rather than as a number.
 *
 * Why a breakdown type instead of a function returning a float. The Engineering
 * page promises that a reader can check the arithmetic by hand, and it was not
 * keeping that promise: it printed a formula string that had been typed out
 * beside the implementation and had drifted from it, and the implementation
 * carried guards -- efficiency forced to zero at zero speed, health not scored
 * at all unless the machine is running -- that appeared nowhere in the printed
 * version. So the page showed
 *
 *     Health = 100 - 12 max(0, RMS - 1.8) - 60 max(0, 0.55 - eta) - ... = 100.0
 *
 * for a stopped machine, where substituting eta = 0 gives 67. The reader who
 * does the substitution concludes the page is lying, and on that page they are
 * the reader who matters most.
 *
 * The fix is not a better string. It is that there is only one artefact: this
 * module computes the terms, frameModel sums them, and Engineering renders the
 * same objects. A term cannot be displayed differently from how it was applied
 * because there is no second place to write it down.
 */

import { ZONE_A_B_MM_S, ZONE_B_C_MM_S, ZONE_C_D_MM_S, vibrationHealthPenalty } from './vibration'
import { ALARM_LIMITS } from './pumpParameters.generated'

/** Below this the suction margin is inside normal engineering practice. */
export const NPSH_MARGIN_GUIDELINE_M = ALARM_LIMITS.npsh_margin_low.alarm
/**
 * Bearing temperature above which grease life starts to be consumed fast.
 * Uses the alarm setpoint from the backend's single source of truth.
 */
export const BEARING_TEMPERATURE_LIMIT_C = ALARM_LIMITS.bearing_temperature_high.alarm
/**
 * Efficiency is penalised below this FRACTION of the BEP efficiency, not below
 * an absolute figure. An absolute threshold silently becomes an always-on
 * penalty the moment the machine's peak efficiency is corrected -- which is
 * exactly what happened when eta_BEP moved from 62 % to 42 %.
 */
export const EFFICIENCY_PENALTY_ONSET_FRACTION = 0.89

export interface HealthTerm {
  label: string
  /** The term as written, with this frame's numbers already substituted. */
  expression: string
  /** Points subtracted. Always >= 0. */
  penalty: number
}

export interface HealthBreakdown {
  /** Every term, including the ones scoring zero -- a hidden zero is a lie. */
  terms: HealthTerm[]
  /** 0..100 after clamping. */
  health: number
  /**
   * False when a guard suppressed scoring entirely. The reader is told which
   * guard and why rather than being shown a 100 with no explanation.
   */
  scored: boolean
  guard: string | null
}

export interface HealthInputs {
  running: boolean
  vibration_rms_mm_s: number
  pump_efficiency: number
  bep_efficiency: number
  npsh_margin_m: number
  bearing_temperature_c: number
}

const f = (value: number, digits = 2) => value.toFixed(digits)

export function healthBreakdown(input: HealthInputs): HealthBreakdown {
  if (!input.running) {
    // Stopped guard, stated rather than hidden. A machine that is not turning
    // has no vibration to judge, no operating point to be efficient at and no
    // suction margin to lose. Scoring it would report a fault-free machine as
    // damaged purely for being switched off.
    return {
      terms: [],
      health: 100,
      scored: false,
      guard: 'Health is not scored unless the asset is RUNNING (stopped guard).',
    }
  }

  const efficiencyFraction =
    input.bep_efficiency > 0 ? input.pump_efficiency / input.bep_efficiency : 0

  const terms: HealthTerm[] = [
    {
      label: 'Vibration',
      expression:
        `piecewise(RMS = ${f(input.vibration_rms_mm_s)} mm/s): ` +
        `4/mm/s to ${ZONE_A_B_MM_S}, 11 to ${ZONE_B_C_MM_S}, 22 to ${ZONE_C_D_MM_S}, 40 beyond`,
      penalty: vibrationHealthPenalty(input.vibration_rms_mm_s),
    },
    {
      label: 'Efficiency',
      expression:
        `37 max(0, ${EFFICIENCY_PENALTY_ONSET_FRACTION} - eta/eta_BEP) = ` +
        `37 max(0, ${EFFICIENCY_PENALTY_ONSET_FRACTION} - ${f(input.pump_efficiency, 3)}/${f(input.bep_efficiency, 2)}) = ` +
        `37 max(0, ${EFFICIENCY_PENALTY_ONSET_FRACTION} - ${f(efficiencyFraction, 3)})`,
      penalty: Math.max(0, (EFFICIENCY_PENALTY_ONSET_FRACTION - efficiencyFraction) * 37),
    },
    {
      label: 'NPSH margin',
      expression: `14 max(0, ${NPSH_MARGIN_GUIDELINE_M} - ${f(input.npsh_margin_m)})`,
      penalty:
        input.npsh_margin_m < NPSH_MARGIN_GUIDELINE_M
          ? (NPSH_MARGIN_GUIDELINE_M - input.npsh_margin_m) * 14
          : 0,
    },
    {
      label: 'Bearing temperature',
      expression: `1.4 max(0, ${f(input.bearing_temperature_c, 1)} - ${BEARING_TEMPERATURE_LIMIT_C})`,
      penalty: Math.max(0, (input.bearing_temperature_c - BEARING_TEMPERATURE_LIMIT_C) * 1.4),
    },
  ]

  const total = terms.reduce((sum, term) => sum + term.penalty, 0)
  return {
    terms,
    // Floored at 5 rather than 0: a machine still running and still reporting
    // is not in a zero-information state, and a bar pinned to the bottom hides
    // the difference between bad and worse.
    health: Math.max(5, Math.min(100, 100 - total)),
    scored: true,
    guard: null,
  }
}

File: backend/app/health.py
"""Visitor-facing health index.

This module is the Python counterpart of `frontend/src/lib/health.ts`. The
front end owns the rendered breakdown, but the backend must score the same
telemetry the same way for API fallback and ML diagnosis payloads.
"""

from __future__ import annotations

from dataclasses import dataclass

from .config.alarm_thresholds import ALARM_LIMITS, VIBRATION
from .config.pump_parameters import PUMP
from .contracts import AssetState, Telemetry

NPSH_MARGIN_GUIDELINE_M = next(limit for limit in ALARM_LIMITS if limit.key == "npsh_margin_low").alarm
BEARING_TEMPERATURE_LIMIT_C = next(
    limit for limit in ALARM_LIMITS if limit.key == "bearing_temperature_high"
).alarm
EFFICIENCY_PENALTY_ONSET_FRACTION = 0.89


@dataclass(frozen=True)
class HealthTerm:
    label: str
    penalty: float


@dataclass(frozen=True)
class HealthBreakdown:
    terms: tuple[HealthTerm, ...]
    health: float
    scored: bool
    guard: str | None


def health_breakdown(telemetry: Telemetry) -> HealthBreakdown:
    """The transparent penalty sum used for visitor-facing health."""
    if telemetry.asset_state != AssetState.RUNNING.value:
        return HealthBreakdown(
            terms=(),
            health=100.0,
            scored=False,
            guard="Health is not scored unless the asset is RUNNING (stopped guard).",
        )

    efficiency_fraction = (
        telemetry.pump_efficiency / PUMP.bep_efficiency if PUMP.bep_efficiency > 0.0 else 0.0
    )
    terms = (
        HealthTerm("Vibration", vibration_health_penalty(telemetry.vibration_rms_mm_s)),
        HealthTerm(
            "Efficiency",
            max(0.0, (EFFICIENCY_PENALTY_ONSET_FRACTION - efficiency_fraction) * 37.0),
        ),
        HealthTerm(
            "NPSH margin",
            (
                (NPSH_MARGIN_GUIDELINE_M - telemetry.npsh_margin_m) * 14.0
                if telemetry.npsh_margin_m < NPSH_MARGIN_GUIDELINE_M
                else 0.0
            ),
        ),
        HealthTerm(
            "Bearing temperature",
            max(0.0, (telemetry.bearing_temperature_c - BEARING_TEMPERATURE_LIMIT_C) * 1.4),
        ),
    )
    total = sum(term.penalty for term in terms)
    return HealthBreakdown(
        terms=terms,
        health=max(5.0, min(100.0, 100.0 - total)),
        scored=True,
        guard=None,
    )


def health_index(telemetry: Telemetry, anomaly_score: float = 0.0) -> float:
    _ = anomaly_score
    return health_breakdown(telemetry).health


def health_terms(telemetry: Telemetry, anomaly_score: float = 0.0) -> dict[str, float]:
    _ = anomaly_score
    return {term.label: term.penalty for term in health_breakdown(telemetry).terms}


def vibration_health_penalty(rms_mm_s: float) -> float:
    rms = max(0.0, float(rms_mm_s))
    in_zone_a = min(rms, VIBRATION.zone_a_b_mm_s)
    in_zone_b = min(max(rms - VIBRATION.zone_a_b_mm_s, 0.0), VIBRATION.zone_b_c_mm_s - VIBRATION.zone_a_b_mm_s)
    in_zone_c = min(max(rms - VIBRATION.zone_b_c_mm_s, 0.0), VIBRATION.zone_c_d_mm_s - VIBRATION.zone_b_c_mm_s)
    in_zone_d = max(rms - VIBRATION.zone_c_d_mm_s, 0.0)
    return 4.0 * in_zone_a + 11.0 * in_zone_b + 22.0 * in_zone_c + 40.0 * in_zone_d


def severity_label(index: float) -> str:
    if index >= 85.0:
        return "none"
    if index >= 65.0:
        return "minor"
    if index >= 40.0:
        return "moderate"
    return "severe"


__all__ = [
    "HealthBreakdown",
    "HealthTerm",
    "health_breakdown",
    "health_index",
    "health_terms",
    "severity_label",
    "vibration_health_penalty",
]
